packages/fxa-auth-server/lib/payments/configuration/manager.ts

```diff
@@ -0,0 +1,177 @@
+  private cancelProductListener: (() => void) | undefined;
+  private cancelPlanListener: (() => void) | undefined;
+
+  private products: Record<string, ProductConfig> = {};
+  private plans: Record<string, PlanConfig> = {};
+
+  constructor() {
+    const config = Container.get(AppConfig);
+    this.prefix = `${config.authFirestore.prefix}payment-config-`;
+    this.firestore = Container.get(AuthFirestore);
+    this.log = Container.get(AuthLogger) as any as Logger;
+    this.productConfigDbRef = this.firestore.collection(
+      `${this.prefix}products`
+    ) as TypedCollectionReference<ProductConfig>;
+    this.planConfigDbRef = this.firestore.collection(
+      `${this.prefix}plans`
+    ) as TypedCollectionReference<PlanConfig>;
+  }
+
+  /**
+   * Load all products and plans from firestore.
+   *
+   * Note that this will overwrite any existing products and plans.
+   */
+  public async load() {
+    const [productResults, planResults] = await Promise.all([
+      this.productConfigDbRef.select().get(),
+      this.planConfigDbRef.select().get(),
+    ]);
+    productResults.docs.forEach((doc) => {
+      this.products[doc.id] = ProductConfig.fromFirestoreObject(
+        doc.data(),
+        doc.id
+      );
+    });
+    planResults.docs.forEach((doc) => {
+      this.plans[doc.id] = PlanConfig.fromFirestoreObject(doc.data(), doc.id);
+    });
+  }
+
+  /**
+   * Start the listener for changes to the product and plan configs.
+   *
+   * Note that this will exit the process if it fails to start the listeners.
+   */
+  public startListener() {
+    this.cancelProductListener = this.productConfigDbRef.onSnapshot(
+      (snapshot) => {
+        snapshot.docChanges().forEach((change) => {
+          if (['added', 'modified'].includes(change.type)) {
+            this.products[change.doc.id] = ProductConfig.fromFirestoreObject(
+              change.doc.data(),
+              change.doc.id
+            );
+          } else {
+            if (change.doc.id in this.products) {
+              delete this.products[change.doc.id];
+            }
+          }
+        });
+      },
+      (err) => {
+        this.log.error('startListener', { err });
+        process.exit(1);
+      }
+    );
+    this.cancelPlanListener = this.planConfigDbRef.onSnapshot((snapshot) => {
+      snapshot.docChanges().forEach(
+        (change) => {
+          if (['added', 'modified'].includes(change.type)) {
+            this.plans[change.doc.id] = PlanConfig.fromFirestoreObject(
+              change.doc.data(),
+              change.doc.id
+            );
+          } else {
+            if (change.doc.id in this.plans) {
+              delete this.plans[change.doc.id];
+            }
+          }
+        },
+        (err) => {
+          this.log.error('startListener', { err });
+          process.exit(1);
+        }
+      );
+    });
+  }
+
+  public stopListener() {
+    this.cancelProductListener?.();
+    this.cancelPlanListener?.();
```

Comment:
Because otherwise this isn't type-correct as these are not necesarilly defined if you haven't started the listener. Omitting the `?` would require a different type-guard or it wouldn't allow you to call a possibly undefined value.